### PR TITLE
hmem, prov/shm: Generic HMEM IPC protocol using get/open/close_handle, implemented those funcs for CUDA

### PR DESCRIPTION
**Pull request by PeterGottesman**
This PR:
 - Adds a new function to the HMEM interface - `is_ipc_enabled`
 - introduces an HMEM IPC protocol into the shm provider based on the `get/open/close_handle` and `is_ipc_enabled` functions in the hmem interface. This uses the existing `smr_progress_ipc` function used by Ze, but as it is not based on dma-buf there is a new `smr_format_ipc` function added.
 - Implements the necessary functions for IPC in the CUDA HMEM interface

The cuda `get/open_handle` functions have a very high overhead, about 750us total, so IPC handle caching would be a very useful addition in the future.

### CODE REVIEW

File: prov/shm/src/smr_msg.c
Comment (aingerson) [inline prov/shm/src/smr_msg.c:228]: This if/else needs to be reworked - ZE will always go to the first case if IPC is enabled (assuming the p2p call gets merged with the IPC enabled call as referenced in a separate comment) and never to the second. We should also be able to rework some of the ZE conditions as well. iov_count can go for sure. ze_ipc_enabled should stay since that has to do with the ZE specific sockets. The other two should probably get added to the use_ipc condition since I think the same conditions apply for both CUDA and ROCR.

File: prov/shm/src/smr_msg.c
Comment (aingerson) [inline prov/shm/src/smr_msg.c:171]: Rename to just `use_ipc` please

File: src/hmem_cuda.c
Comment (PeterGottesman) [reply src/hmem_cuda.c:268]: Okay, I will nix the exit in the next revision. This function has a well-defined prototype (that of cudaMemcpy) so it should return a cudaError_t.

File: src/hmem_cuda.c
Comment (PeterGottesman) [reply src/hmem_cuda.c:448]: This is used as a parameter to `fi_param_get_bool`, which takes an `int *` as the argument.

File: include/ofi_hmem.h
Comment (PeterGottesman) [reply include/ofi_hmem.h:106]: `is_ipc_enabled` is used in this PR to signify that IPC can be performed using exclusively the IPC functions defined in the HMEM interface. My understanding is that Ze needs extra logic on top of that interface, so I did not think it made sense to have `is_ipc_enabled` return true in any case for it. Am I wrong in that regard?

Comment (aingerson) [reply include/ofi_hmem.h:106]: Ze needs extra logic in the sense that the device fds need to be exchanged via a socket instead of directly through the shared memory but it still uses the get/open/close handle functions. The p2p check just checks whether the devices can perform an IPC call across the devices. It should be fairly similar to what the CUDA and ROCR versions are checking.

File: src/hmem_cuda.c
Comment (aingerson) [reply src/hmem_cuda.c:448]: Sorry yes that part is fine. I meant to save it for later. This one looks like it's local so it's fine. The other one below is global so that's why I think it makes more sense as a bool but it's not a big deal. The other providers do something like this:
```
        fi_param_get_bool(&rxm_prov, "enable_dyn_rbuf", &ret);
        domain->dyn_rbuf = (ret != 0);
```

File: src/hmem_cuda.c
Comment (PeterGottesman) [reply src/hmem_cuda.c:264]: It does, but I don't think there is anything here that really should be cut. This error message should communicate what went wrong and should be done. If it needs to be changed I guess I can slim it down, maybe:
```C
		"cudaMemcpy was called but FI_HMEM_CUDA_ENABLE_XFER = 0, "
		"no copy will occur. You may want to enable gdrcopy with FI_HMEM_CUDA_USE_GDRCOPY. "
		"The provider in use may not support disabling cudaMemcpy, if you are unsure you can reach out to "
		"the libfabric developers via a github issue.");
```

File: include/ofi_hmem.h
Comment (PeterGottesman) [reply include/ofi_hmem.h:106]: Yes, they will all have some check for IPC capability, but my intention was for this function to be about checking if the interface can use this "generic IPC" protocol, i.e. IPC can be done without any interface specific codepaths. If interface specific code is needed, then the function should return false and leave it up to the interface to add the proper conditionals in shm protocol selection. I can make this return true for Ze, but then we are just making protocol selection evaluate more conditionals to get the same effect, especially if more HMEM interfaces need their own special codepath. 
Maybe I am looking at this wrong if CUDA is the odd one out in that it doesn't need extra logic for IPC. If that is the  maybe the HMEM interface should be modified to support the typical use case?

Comment (aingerson) [reply include/ofi_hmem.h:106]: I think my problem with this is there's not really a "generic IPC" protocol. It will inherently depend on the underlying interface. CUDA does it one way. ZE does it another way.
Ze needs some interface specific things to make it work which shm does check for and take into account as you suggest but it still uses the "generic" IPC get/open/close handle calls.
All you're doing in the CUDA check for IPC is allocating memory and getting a pointer so by that definition, the same check in ZE would pass. There is nothing in the CUDA IPC check that ZE would fail in the ZE world so I don't think that's enough to say ZE should fail.
Since CUDA was the first major device API to come out it's easy to say "well this API doesn't match up exactly with CUDA so they are the odd ones out" but I think they're just different and require different things. As far as your function that asks "does this interface support IPC?" I think ZE is ok to pass that and have a separate ZE-specific check in shm to see if shm is doing the special extra things to support ZE IPC.

Comment (PeterGottesman) [reply include/ofi_hmem.h:106]: That all sounds reasonable to me. I will make these changes now.

File: src/hmem_cuda.c
Comment (rwespetal) [reply src/hmem_cuda.c:264]: Suggest "no copy will occur" -> "no copy will occur to prevent deadlock."

I don't understand the second sentence or what a user would do with that info.

File: src/hmem_cuda.c
Comment (rwespetal) [inline src/hmem_cuda.c:469]: I probably wouldn't refer to a specific section as this will likely go stale. I'd also suggest that these be trimmed down and this info is added into a man page somewhere.

File: src/hmem_cuda.c
Comment (rwespetal) [inline src/hmem_cuda.c:481]: Since this sets up a thread and does some other init the param should be checked first so the init can be skipped when gdrcopy is disabled.

File: prov/shm/src/smr_ep.c
Comment (rwespetal) [inline prov/shm/src/smr_ep.c:404]: I wouldn't take an iovec as a param to this function if it doesn't support more than one.

File: src/hmem_cuda.c
Comment (PeterGottesman) [reply src/hmem_cuda.c:469]: There is not currently any man page for HMEM. Should I add one, or append this information to a different man page?

Comment (rwespetal) [reply src/hmem_cuda.c:469]: I'm fine with it landing in an existing man page if there's a good spot for it.

**Comment by PeterGottesman** [comment]
@aingerson Would it be possible for you to review my changes today? Specifically the changes which affect Ze, as I cannot test that AFAIK.

**Comment by aingerson** [comment]
@PeterGottesman Yeah that's my plan for today. Our systems are down right now but as soon as they come back up this is my priority. Code looks fine to me. I'll approve once I verify on our systems.

**Review by rwespetal — APPROVED** [review]

**Review by aingerson — APPROVED** [review]
Have confirmed that these changes work with ze though a fabtests bug fix is needed to invoke the IPC path. I'll fix that separately.
I'm assuming you've confirmed this works with CUDA? I noticed that fabtests does not currently have a way to test specific CUDA devices and the same fabtests bug for ZE would affect the CUDA path so just want to make sure you have a validation vehicle.

**Comment by PeterGottesman** [comment]
Yup, this was validated with CUDA using OMB + a modified EFA provider, I did not run tests where I selected specific CUDA devices though. What was the ZE bug in fabtests?

**Comment by aingerson** [comment]
Since IPC is only valid for device only allocations, we have to register with the FI_HMEM_DEVICE_ONLY flag in order to properly get to the IPC path:
```
diff --git a/fabtests/common/shared.c b/fabtests/common/shared.c
index 583ee001a..5bd6440b8 100644
--- a/fabtests/common/shared.c
+++ b/fabtests/common/shared.c
@@ -392,7 +392,8 @@ static int ft_reg_mr(void *buf, size_t size, uint64_t access,
                break;
        }

-       ret = fi_mr_regattr(domain, &attr, 0, mr);
+       ret = fi_mr_regattr(domain, &attr,
+                           opts.iface ? FI_HMEM_DEVICE_ONLY : 0, mr);
        if (ret)
                return ret;
```

**Comment by PeterGottesman** [comment]
Got it, that would have been covered by the modified EFA provider in my testing. I don't believe more testing is needed here as a result of that issue, so it should be good to merge.

**Pull request merged**
